### PR TITLE
LangChain Community: VectorStores: Azure Cosmos DB Filtered Vector Search

### PR DESCRIPTION
**Pull request by aayush3011**
Thank you for contributing to LangChain!

- This PR adds vector search filtering for Azure Cosmos DB Mongo vCore and NoSQL.

- [ ] **PR message**: ***Delete this entire checklist*** and replace with
    - **Description:** a description of the change
    - **Issue:** the issue # it fixes, if applicable
    - **Dependencies:** any dependencies required for this change
    - **Twitter handle:** if your PR gets announced, and you'd like a mention, we'll gladly shout you out!

- [ ] **Add tests and docs**: If you're adding a new integration, please include
  1. a test for the integration, preferably unit tests that do not rely on network access,
  2. an example notebook showing its use. It lives in `docs/docs/integrations` directory.

- [ ] **Lint and test**: Run `make format`, `make lint` and `make test` from the root of the package(s) you've modified. See contribution guidelines for more: https://python.langchain.com/docs/contributing/

Additional guidelines:
- Make sure optional dependencies are imported within a function.
- Please do not add dependencies to pyproject.toml files (even optional ones) unless they are required for unit tests.
- Most PRs should not touch more than one package.
- Changes should be backwards compatible.
- If you are adding something to community, do not re-import it in langchain.

If no one reviews your PR within a few days, please @-mention one of baskaryan, efriis, eyurtsev, ccurme, vbarda, hwchase17.

### CODE REVIEW

File: libs/community/langchain_community/vectorstores/azure_cosmos_db.py [hunk 1]
```diff
@@ -397,8 +418,10 @@ def _similarity_search_with_score(
         embeddings: List[float],
         k: int = 4,
         kind: CosmosDBVectorSearchType = CosmosDBVectorSearchType.VECTOR_IVF,
+        pre_filter: Optional[Dict] = None,
         ef_search: int = 40,
         score_threshold: float = 0.0,
+        with_embedding: bool = False,
```
Comment (isahers1) [inline libs/community/langchain_community/vectorstores/azure_cosmos_db.py:424]: Can you add new args to docstring

File: libs/community/tests/integration_tests/vectorstores/test_azure_cosmos_db.py [hunk 2]
```diff
@@ -904,6 +904,111 @@ def test_max_marginal_relevance_inner_product_vector_hnsw(
...
+    output = vectorstore.similarity_search(
+        "Dogs",
+        k=4,
+        kind=kind,
+        ef_search=ef_search,
+        score_threshold=score_threshold,
+        pre_filter={
+            "metadata.a": {
+                "$eq": 5
+            }
```
Comment (isahers1) [inline libs/community/tests/integration_tests/vectorstores/test_azure_cosmos_db.py:947]: Should `Document(page_content="Dogs are tough.", metadata={"a": 1}),` have "a":5?

File: libs/community/langchain_community/vectorstores/azure_cosmos_db.py (hunk 1 above)
Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db.py:424]: @isahers1  What do you mean by docstring?

File: libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py [hunk 3]
```diff
@@ -257,54 +264,97 @@ def _similarity_search_with_score(
...
-                self._embedding_key,
-                embeddings,
-                self._embedding_key,
-                embeddings,
-            )
+        query = "SELECT "
+
+        # If limit_offset_clause is not specified, add TOP clause
+        if pre_filter is None or pre_filter.get("limit_offset_clause") is None:
+            query += "TOP {} ".format(k)
```
[the remark this answers is not in the document]
Comment (aayush3011) [reply libs/community/langchain_community/vectorstores/azure_cosmos_db_no_sql.py:274]: For CosmosDB NoSQL vector search, Limit clause will always have both offset and limit in it. We don't support offset without limit. I will add this statement in the readme as well.

File: libs/community/tests/integration_tests/vectorstores/test_azure_cosmos_db.py (hunk 2 above)
Comment (aayush3011) [reply libs/community/tests/integration_tests/vectorstores/test_azure_cosmos_db.py]: Fixed the test case

**Comment by isahers1** [comment]
@aayush3011 Can you run make lint and make format to fix the listing errors, and then I think it should be good to merge

**Pull request merged**